### ISSUE TITLE
gpfdist does not work on mac due to newest curl incompatibility with old version

### ISSUE THREAD

**Issue by yydzero** [greenplum-db/gpdb#4105]
external table code is based on curl, extracted core code from Greenplum's url_curl.c as below, these same code works on RHEL 6 & 7, while does not work on MacOS Sierra.  This breaks gpfdist support on mac os.

On Mac: response_code should be 200, while it is 0.
```
± |gpfdist-kill {2} U:1 ?:10 ✗| → ./a.out
responde_code = 0
* Curl_http_done: called premature == 1
* Closing connection 0
```

RHEL 7.2 curl version is 7.29.0
MacOS Sierra has used a much newer version of libcurl (7.51.0)

```
#include <stdio.h>
#include <curl/curl.h>

int
main(int argc, char **argv) {
	CURL *curl = curl_easy_init();

	static char curl_Error_Buffer[CURL_ERROR_SIZE];

	curl_Error_Buffer[0] = '\0';

	if (!curl) {
		printf("failed to init curl\n");
	}

	CURLcode res;

	curl_easy_setopt(curl, CURLOPT_URL, "http://localhost:7070/a.txt");
	if (res != CURLE_OK)
		printf("failed to setopt\n");
	curl_easy_setopt(curl, CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);
	if (res != CURLE_OK)
		printf("failed to setopt\n");
	curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_Error_Buffer);
	if (res != CURLE_OK)
		printf("failed to setopt\n");

//        curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, CURL_IPRESOLVE_V4);
//        curl_easy_setopt(curl, CURLOPT_WRITEHEADER, CURL_IPRESOLVE_V4);
//        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, CURL_IPRESOLVE_V4);
//        curl_easy_setopt(curl, CURLOPT_WRITEDATA, file);


	curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
	if (res != CURLE_OK)
		printf("failed to setopt\n");

	struct curl_slist *chunk = NULL;
	/* Remove a header curl would otherwise add by itself */
	chunk = curl_slist_append(chunk, "X-GP-PROTO:1");
	/* set our custom set of headers */
	res = curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
	if (res != CURLE_OK)
		printf("failed to setopt\n");


	CURLM *multi_handle = 0;

	if (!(multi_handle = curl_multi_init()))
		printf("FAILED to create multiple handle\n");


	if (CURLE_OK != (res = curl_multi_add_handle(multi_handle, curl))) {
		printf("FAILED to create multiple handle\n");
	}
	int still_running;
	CURLMcode e;

	while (CURLM_CALL_MULTI_PERFORM ==
		   (e = curl_multi_perform(multi_handle, &still_running)));

	//fill_buffer(file, 1);

	if (e != CURLE_OK) {
		printf("failed to perf multi\n");
	}

	long response_code;
	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
	printf("responde_code = %ld\n", response_code);

	if (curl_Error_Buffer[0] != '\0')
		printf("error: %s\n", curl_Error_Buffer);


//	  	res = curl_easy_perform(curl);
//		if (res == CURLE_OK) {
//			long response_code;
//			curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
//			printf("responde_code = %ld\n", response_code);
//		}

	curl_easy_cleanup(curl);

	curl_multi_cleanup(multi_handle);
}

```

**Comment by yydzero** [comment]
@jchampio FYI, gpfdist on mac does not work now.

**Comment by lavjain** [comment]
@yydzero Please take a look at https://github.com/lavjain/gpdb/commit/f12d756c4e096f320455fe130bd24cf159352db0#diff-a1c0c1aa48f9ec08937199b5bfbda24b where we fixed a similar issue for OS X Sierra.

**Comment by yydzero** [comment]
Thanks @lavjain Could you please explain a bit more about why the changes needed in fill_internal_buffer()?  looks like the major change is to call curl_multi_timeout().

**Comment by lavjain** [comment]
Please take a look at the examples given by curl website. The documented way of doing the curl calls is documented. (The old code might have worked at some point in the past)

https://curl.haxx.se/libcurl/c/multi-app.html

**Comment by Eulerizeit** [comment]
literally this is a RTFM email.

**Comment by adam8157** [comment]
Fixed by #10261
